### ISSUE TITLE
Compile error under VS2015 Update 3 with `/std:c++latest`

### ISSUE THREAD

**Issue by LYP951018** [philsquared/Catch#697]
With Visual Studio 2015 Update 3, /std:c++latest removes `std::random_shuffle`. Unfortunately, the switch between the usage of `std::random_shuffle` and `std::shuffle` depends on the value of `__cplusplus`, which is 199711. Fix:

``` cpp
#ifdef CATCH_CPP14_OR_GREATER
#define USE_MODERN_SHUFFLE
#else
#ifdef _HAS_AUTO_PTR_ETC
#if !_HAS_AUTO_PTR_ETC
#define USE_MODERN_SHUFFLE
#endif
#endif
#endif
```

**Comment by polyvertex** [comment]
For the record and to add to @LYP951018's comment, the `__cplusplus` value with MSVC is [not reliable](https://connect.microsoft.com/VisualStudio/feedback/details/763051/a-value-of-predefined-macro-cplusplus-is-still-199711l). At least not the way Catch expects it to be.

**Comment by horenmar** [comment]
This should be fixed by 0016273dbf6f910e547c6274360bf297ad6ff5ec.

**Comment by horenmar** [comment]
The commit is now in master, so it should be solved (reopen if it isn't).
